diff -uNr ./kdelibs-3.96.0.org/kate/document/katedocument.cpp ./kdelibs-3.96.0/kate/document/katedocument.cpp
--- ./kdelibs-3.96.0.org/kate/document/katedocument.cpp	2007-11-14 07:18:17.000000000 +0700
+++ ./kdelibs-3.96.0/kate/document/katedocument.cpp	2008-01-02 18:36:47.000000000 +0700
@@ -561,6 +561,16 @@
   return l->string();
 }
 
+bool KateDocument::getBreakLine(int line, int col)
+{
+  KateTextLine::Ptr l = m_buffer->plainLine(line);
+
+  if(!l)
+    return 0;
+  
+  return l->isBreakable(col);
+}
+
 bool KateDocument::setText(const QString &s)
 {
   if (!isReadWrite())
@@ -1193,13 +1203,15 @@
       for (z=searchStart; z > 0; z--)
       {
         if (t.at(z).isSpace()) break;
-        if ( ! nw && highlight()->canBreakAt( t.at(z) , l->attribute(z) ) )
+        if (l->isBreakable(z)) break;
+        if ( ! nw && highlight()->canBreakAt( t.at(z) , l->attribute(z) ))
         nw = z;
       }
 
       if (z > 0)
       {
         // cu space
+        if (!l->isBreakable(z))
         editRemoveText (line, z, 1);
       }
       else
diff -uNr ./kdelibs-3.96.0.org/kate/document/katedocument.h ./kdelibs-3.96.0/kate/document/katedocument.h
--- ./kdelibs-3.96.0.org/kate/document/katedocument.h	2007-11-14 07:18:17.000000000 +0700
+++ ./kdelibs-3.96.0/kate/document/katedocument.h	2008-01-02 17:57:11.000000000 +0700
@@ -182,6 +182,7 @@
     virtual QString line(int line) const;
     virtual QChar character(const KTextEditor::Cursor& position) const;
     int lines() const;
+    bool getBreakLine(int line, int col);
     virtual KTextEditor::Cursor documentEnd() const;
     int numVisLines() const;
     int totalCharacters() const;
diff -uNr ./kdelibs-3.96.0.org/kate/document/katetextline.cpp ./kdelibs-3.96.0/kate/document/katetextline.cpp
--- ./kdelibs-3.96.0.org/kate/document/katetextline.cpp	2007-11-14 07:18:17.000000000 +0700
+++ ./kdelibs-3.96.0/kate/document/katetextline.cpp	2008-01-02 18:07:07.000000000 +0700
@@ -28,6 +28,107 @@
 
 #include <QtCore/QRegExp>
 
+//Load Libthai
+#include <QtCore/QTextCodec>
+#include <qlibrary.h>
+#ifndef HAVE_LIBTHAI
+typedef int (*th_brk_def)(const unsigned char*, int[], int);
+static th_brk_def th_brk;
+#else
+#include <thai/thailib.h>
+#include <thai/thbrk.h>
+#endif
+
+//Thai Break
+struct ThaiCache
+{
+    ThaiCache() {
+        string = 0;
+        allocated = 0x400;
+        wbrpos = (int *) malloc(allocated*sizeof(int));
+        numwbrpos = 0;
+        numisbreakable = 0x400;
+        isbreakable = (int *) malloc(numisbreakable*sizeof(int));
+    }
+    ~ThaiCache() {
+        free(wbrpos);
+        free(isbreakable);
+    }
+    const QChar *string;
+    int *wbrpos;
+    int *isbreakable;
+    int allocated;
+    int numwbrpos,numisbreakable;
+};
+static ThaiCache *cache = 0;
+
+void cleanup_thaibreaks()
+{
+    delete cache;
+    cache = 0;
+#ifndef HAVE_LIBTHAI
+    th_brk = 0;
+#endif
+}
+
+int* breakWordThai( const QChar *string, const int pos, const int len)
+{
+    static QTextCodec *thaiCodec = QTextCodec::codecForMib(2259);
+
+#ifndef HAVE_LIBTHAI
+    // KLibLoader requires libthai.la which is usually in libthai-dev package or doesn't exist at all
+    // QLibrary::resolve will look for libthai.so.0 directly
+    /* load libthai dynamically */
+    if (!th_brk && thaiCodec) {
+        printf("Try to load libthai dynamically...\n");
+        th_brk = (th_brk_def)QLibrary::resolve("thai", "th_brk");
+        if (!th_brk) {
+        printf("Error, can't load libthai...\n");
+            thaiCodec = 0;
+        }
+    }
+
+    if (!th_brk ) {
+        return NULL;
+    }
+#endif
+
+    if (!cache ) {
+        cache = new ThaiCache;
+#ifndef HAVE_LIBTHAI
+            //cache->library = lib;
+#endif
+    }
+    // build up string of thai chars
+    if ( string != cache->string ) {
+        //fprintf(stderr,"new string found (not in cache), calling libthai\n");
+        QByteArray cstr = thaiCodec->fromUnicode( QString::fromRawData(string,len));
+        //printf("About to call libthai::th_brk with str: %s",cstr.data());
+         cache->numwbrpos = th_brk((const unsigned char*) cstr.data(), cache->wbrpos, cache->allocated);
+        //fprintf(stderr,"libthai returns with value %d\n",cache->numwbrpos);
+        if (cache->numwbrpos > cache->allocated) {
+            cache->allocated = cache->numwbrpos;
+            cache->wbrpos = (int *)realloc(cache->wbrpos, cache->allocated*sizeof(int));
+            cache->numwbrpos = th_brk((const unsigned char*) cstr.data(), cache->wbrpos, cache->allocated);
+        }
+        if ( len > cache->numisbreakable ) {
+            cache->numisbreakable=len;
+            cache->isbreakable = (int *)realloc(cache->isbreakable, cache->numisbreakable*sizeof(int));
+        }
+        for (int i = 0 ; i < len ; ++i) {
+            cache->isbreakable[i] = 0;
+        }
+        if ( cache->numwbrpos > 0 ) {
+            for (int i = cache->numwbrpos-1; i >= 0; --i) {
+                cache->isbreakable[cache->wbrpos[i]] = 1;
+            }
+        }
+        cache->string = string;
+    }
+ 
+    return cache->isbreakable;
+}
+
 KateTextLine::KateTextLine ()
   : m_flags(0)
 {
@@ -49,6 +150,17 @@
     return;
 
   m_text.insert (pos, insText);
+  
+  // Thai word break
+  m_breakLine.resize (m_text.length());  
+  int* bw_thai = breakWordThai(m_text.constData(),0,m_text.length());
+  for (int i = 0;i < m_text.length(); i++)
+    m_breakLine.insert (i,bw_thai[i]); 
+}
+
+bool KateTextLine::isBreakable (int pos)
+{
+  return m_breakLine[pos];
 }
 
 void KateTextLine::removeText (uint pos, uint delLen)
@@ -69,6 +181,7 @@
     delLen = textLen - pos;
 
   m_text.remove (pos, delLen);
+  m_breakLine.remove (pos, delLen);
 }
 
 void KateTextLine::truncate(int newLen)
diff -uNr ./kdelibs-3.96.0.org/kate/document/katetextline.h ./kdelibs-3.96.0/kate/document/katetextline.h
--- ./kdelibs-3.96.0.org/kate/document/katetextline.h	2007-07-26 17:39:55.000000000 +0700
+++ ./kdelibs-3.96.0/kate/document/katetextline.h	2008-01-02 17:59:25.000000000 +0700
@@ -285,6 +285,12 @@
     void insertText (int pos, const QString& insText);
 
     /**
+     * return breakline attribute m_breakLine
+     * @param pos insert position
+     */    
+    bool isBreakable (int pos);
+
+    /**
      * remove text at given position
      * @param pos start position of remove
      * @param delLen length to remove
@@ -375,6 +381,11 @@
      */
     QVector<int> m_attributesList;
 
+     /**
+     * Attribute for lineBreak value "true" for breakline
+     */
+    QVector<bool> m_breakLine;
+
     /**
      * context stack
      */
diff -uNr ./kdelibs-3.96.0.org/kate/view/kateviewinternal.cpp ./kdelibs-3.96.0/kate/view/kateviewinternal.cpp
--- ./kdelibs-3.96.0.org/kate/view/kateviewinternal.cpp	2007-11-06 14:59:11.000000000 +0700
+++ ./kdelibs-3.96.0/kate/view/kateviewinternal.cpp	2008-01-02 18:04:20.000000000 +0700
@@ -715,7 +715,11 @@
 
 void KateViewInternal::doDelete()
 {
-  m_doc->del( m_view, m_cursor );
+  if ( m_view->config()->persistentSelection() || !m_view->selection() ) {
+    m_view->shiftCursorRight();
+  }
+  m_view->removeSelectedText();
+  return;
 }
 
 void KateViewInternal::doBackspace()
@@ -1007,8 +1011,11 @@
   }
   else if( h->isInWord( m_doc->line( c.line() )[ c.column() - 1 ] ) )
   {
-    while( !c.atEdge( left ) && h->isInWord( m_doc->line( c.line() )[ c.column() - 1 ] ) )
-      --c;
+    if (m_doc->getBreakLine(c.line(),c.column()))
+       --c;
+
+    while( !c.atEdge( left ) && h->isInWord( m_doc->line( c.line() )[ c.column() - 1 ] ) && !m_doc->getBreakLine(c.line(),c.column()))
+       --c;
   }
   else
   {
@@ -1045,7 +1052,9 @@
   }
   else if( h->isInWord( m_doc->line( c.line() )[ c.column() ] ) )
   {
-    while( !c.atEdge( right ) && h->isInWord( m_doc->line( c.line() )[ c.column() ] ) )
+    if (m_doc->getBreakLine(c.line(),c.column()))
+      ++c;
+    while( !c.atEdge( right ) && h->isInWord( m_doc->line( c.line() )[ c.column() ] ) && !m_doc->getBreakLine(c.line(),c.column()) )
       ++c;
   }
   else
@@ -3373,7 +3382,7 @@
       return renderer()->currentFont();
 
     case Qt::ImCursorPosition:
-      return cursorCoordinates(false);
+      return m_displayCursor.column();
 
     case Qt::ImSurroundingText:
       if (KateTextLine::Ptr l = textLine(m_cursor.line()))
@@ -3399,7 +3408,7 @@
   }
 
   // Finished this input method context?
-  if (m_imPreedit && e->preeditString().isEmpty() && e->commitString().isEmpty()) {
+  if (m_imPreedit && e->preeditString().isEmpty() && e->commitString().isEmpty() && e->replacementLength() == 0) {
     m_view->removeInternalHighlight(m_imPreedit);
     delete m_imPreedit;
     m_imPreedit = 0L;
@@ -3421,7 +3430,7 @@
   if (!e->commitString().isEmpty() || e->replacementLength()) {
     KTextEditor::Range preeditRange = *m_imPreedit;
 
-    KTextEditor::Cursor start(m_imPreedit->start().line(), m_imPreedit->start().column() + e->replacementStart());
+    KTextEditor::Cursor start(m_cursor.line(), m_cursor.column() + e->replacementStart());
     KTextEditor::Cursor removeEnd = start + KTextEditor::Cursor(0, e->replacementLength());
 
     m_view->doc()->editStart(true);
