diff -uNr kdelibs-3.96.0.org/kate/view/kateviewinternal.cpp kdelibs-3.96.0/kate/view/kateviewinternal.cpp
--- kdelibs-3.96.0.org/kate/view/kateviewinternal.cpp	2007-11-06 14:59:11.000000000 +0700
+++ kdelibs-3.96.0/kate/view/kateviewinternal.cpp	2007-12-11 13:43:23.000000000 +0700
@@ -715,7 +715,12 @@
 
 void KateViewInternal::doDelete()
 {
-  m_doc->del( m_view, m_cursor );
+//  m_doc->del( m_view, m_cursor );
+  if ( m_view->config()->persistentSelection() || !m_view->selection() ) {
+    m_view->shiftCursorRight();
+  }
+  m_view->removeSelectedText();
+  return;
 }
 
 void KateViewInternal::doBackspace()
@@ -3373,7 +3378,8 @@
       return renderer()->currentFont();
 
     case Qt::ImCursorPosition:
-      return cursorCoordinates(false);
+//      return cursorCoordinates(false);
+      return m_displayCursor.column();
 
     case Qt::ImSurroundingText:
       if (KateTextLine::Ptr l = textLine(m_cursor.line()))
@@ -3399,7 +3405,8 @@
   }
 
   // Finished this input method context?
-  if (m_imPreedit && e->preeditString().isEmpty() && e->commitString().isEmpty()) {
+//  if (m_imPreedit && e->preeditString().isEmpty() && e->commitString().isEmpty()) {
+  if (m_imPreedit && e->preeditString().isEmpty() && e->commitString().isEmpty() && e->replacementLength() == 0) {
     m_view->removeInternalHighlight(m_imPreedit);
     delete m_imPreedit;
     m_imPreedit = 0L;
@@ -3421,7 +3428,8 @@
   if (!e->commitString().isEmpty() || e->replacementLength()) {
     KTextEditor::Range preeditRange = *m_imPreedit;
 
-    KTextEditor::Cursor start(m_imPreedit->start().line(), m_imPreedit->start().column() + e->replacementStart());
+//    KTextEditor::Cursor start(m_imPreedit->start().line(), m_imPreedit->start().column() + e->replacementStart());
+    KTextEditor::Cursor start(m_cursor.line(), m_cursor.column() + e->replacementStart());
     KTextEditor::Cursor removeEnd = start + KTextEditor::Cursor(0, e->replacementLength());
 
     m_view->doc()->editStart(true);
