diff -uNr kdelibs-4.0.0.orig/khtml/rendering/break_lines.cpp kdelibs-4.0.0/khtml/rendering/break_lines.cpp
--- kdelibs-4.0.0.orig/khtml/rendering/break_lines.cpp	2008-01-17 10:29:29.000000000 +0700
+++ kdelibs-4.0.0/khtml/rendering/break_lines.cpp	2008-01-18 21:24:15.000000000 +0700
@@ -10,7 +10,7 @@
  * libthai since compile time. Otherwise it will try to
  * dlopen at run-time
  *
- * Ott Pattara Nov 14, 2004
+ * Ott Pattara, Jakkapun Kwanroengjai Jan 17, 2008
  */
 
 #ifndef HAVE_LIBTHAI
@@ -101,8 +101,8 @@
 
             cache->numwbrpos = th_brk((const unsigned char*) cstr.data(), cache->wbrpos, cache->allocated);
             //fprintf(stderr,"libthai returns with value %d\n",cache->numwbrpos);
-            if (cache->numwbrpos > cache->allocated) {
-                cache->allocated = cache->numwbrpos;
+            if (cache->numwbrpos >= cache->allocated) {
+                cache->allocated = len;
                 cache->wbrpos = (int *)realloc(cache->wbrpos, cache->allocated*sizeof(int));
                 cache->numwbrpos = th_brk((const unsigned char*) cstr.data(), cache->wbrpos, cache->allocated);
             }
