diff -uNr khtml-oldpatch/rendering/break_lines.cpp khtml/rendering/break_lines.cpp
--- khtml-oldpatch/rendering/break_lines.cpp	2005-03-01 17:47:22.000000000 +0900
+++ khtml/rendering/break_lines.cpp	2005-03-01 17:46:20.000000000 +0900
@@ -15,7 +15,7 @@
 
 #ifndef HAVE_LIBTHAI
 typedef int (*th_brk_def)(const unsigned char*, int[], int);
-static th_brk_def th_brk;
+static th_brk_def th_brk = 0;
 #else
 #include <thai/thailib.h>
 #include <thai/thbrk.h>
@@ -50,6 +50,10 @@
     void cleanup_thaibreaks()
     {
         delete cache;
+	cache = 0;
+#ifndef HAVE_LIBTHAI
+	th_brk = 0;
+#endif
     }
 
     bool isBreakableThai( const QChar *string, const int pos, const int len)
